### PR TITLE
[Serve] Fail early for user app failure and expose failure reasons

### PR DESCRIPTION
**Pull request by Michaelvll**
Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [x] All smoke tests: `pytest tests/test_smoke.py` 
  - [x] `pytest tests/test_smoke.py --serve`
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
  - [x] `pytest tests/test_smoke.py::test_skyserve_user_bug_restart`
  - [x] `pytest tests/test_smoke.py::test_skyserve_failures`
  - [x] `pytest tests/test_smoke.py --serve`
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

**Review by cblmemo — COMMENTED** [review]
Thanks! It looks mostly great for me. Left some nits and after all smoke tests added and passed it should be good to go 🫡

File: tests/skyserve/spot/base_ondemand_fallback.yaml [hunk 1]
```diff
@@ -6,7 +6,6 @@ service:
     min_replicas: 2
     max_replicas: 3
     base_ondemand_fallback_replicas: 1
-    # Use a large qps per replica to avoid scale up for testing purpose.
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply tests/skyserve/spot/base_ondemand_fallback.yaml]: Ahh, it is a merging issue. Fixed.

File: sky/serve/autoscalers.py [hunk 2]
```diff
@@ -382,6 +395,15 @@ def evaluate_scaling(
...
                     latest_nonterminal_replicas.append(info)
+                    if info.is_ready:
+                        self.latest_version_ever_ready = self.latest_version
+                elif (info.status_property.unrecoverable_failure() and
+                      self.latest_version_ever_ready < self.latest_version):
+                    # Stop scaling if one of replica of the latest version
+                    # failed, it is likely that a fatal error happens to the
+                    # user application and may lead to a infinte termination
+                    # and restart.
+                    return []
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/serve/autoscalers.py]: Good point! When `unrecoverable_failure()` happens, it is likely all the replicas will fail, but changing it to two loops should be safer.

File: sky/serve/autoscalers.py [hunk 3]
```diff
@@ -113,13 +114,25 @@ def from_spec(cls, service_name: str,
         else:
             return RequestRateAutoscaler(service_name, spec)
 
+    def _dump_dynamic_states(self) -> Dict[str, Any]:
+        """Dump dynamic states from autoscaler."""
+        raise NotImplementedError
+
     def dump_dynamic_states(self) -> Dict[str, Any]:
         """Dump dynamic states from autoscaler."""
+        states = {'latset_version_ever_ready': self.latest_version_ever_ready}
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/serve/autoscalers.py]: Good catch! Fixed. Thanks!

File: sky/serve/autoscalers.py [hunk 4]
```diff
@@ -75,6 +75,7 @@ def __init__(self, service_name: str,
         # Target number of replicas is initialized to min replicas
         self.target_num_replicas: int = spec.min_replicas
         self.latest_version: int = constants.INITIAL_VERSION
+        self.latest_version_ever_ready: int = self.latest_version - 1
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/serve/autoscalers.py:81]: Added the comment:
```
# The latest_version_ever_ready should be smaller than the
# latest_version, so we can fail early if the inital version got
# unrecoverable failure.
```

File: sky/serve/replica_managers.py [hunk 5]
```diff
@@ -333,17 +340,17 @@ def to_replica_status(self) -> serve_state.ReplicaStatus:
                 return serve_state.ReplicaStatus.FAILED
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/serve/replica_managers.py:340]: I am thinking to use `FAILED` to always mean `FAILED_USER_APP`, as `FAILED` represents user app error in managed spot job. Wdyt? cc'ing @concretevitamin @romilbhardwaj for some input here as well : )

File: sky/serve/replica_managers.py [hunk 6]
```diff
@@ -235,52 +235,59 @@ class ReplicaStatusProperty:
...
+                # user code and the scale down is successful, thus enabling the
+                # controller to remove the replica from the replica table and
+                # auto restart the replica.
+                # For replica with a failed sky.launch, it is likely due to some
+                # misconfigured resources, so we don't want to auto restart it.
+                # For replica with a failed sky.down, we cannot restart it since
+                # otherwise we will have a resource leak.
+                return False
+            elif self.first_ready_time < 0:
+                # If the service timeout, it is likely the service is not
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/serve/replica_managers.py]: Good catch! Updated. Thanks!

File: sky/serve/serve_state.py [hunk 7]
```diff
@@ -98,9 +98,18 @@ class ReplicaStatus(enum.Enum):
     # The replica VM is being shut down. i.e., the `sky down` is still running.
     SHUTTING_DOWN = 'SHUTTING_DOWN'
 
-    # The replica VM is once failed and has been deleted.
+    # The replica fails due to user's run/setup.
     FAILED = 'FAILED'
 
+    # The replica fails due to initial delay.
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/serve/serve_state.py:104]: Fixed. Thanks!

File: tests/skyserve/failures/probing.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='SkyServe HTTP Test Server')
+    parser.add_argument('--port', type=int, required=False, default=8081)
+    args = parser.parse_args()
+
+    Handler = MyHttpRequestHandler
+    with socketserver.TCPServer(('', args.port), Handler) as httpd:
+        print('serving at port', args.port)
+        for i in range(20):
+            # Ends after handle 10 requests
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply tests/skyserve/failures/probing.py:37]: Fixed. Thanks!

File: tests/test_smoke.py [hunk 9]
```diff
@@ -3031,9 +3031,11 @@ def _get_service_name() -> str:
 # The teardown command has a 10-mins timeout, so we don't need to do
 # the timeout here. See implementation of run_one_test() for details.
 _TEARDOWN_SERVICE = (
-    '(while true; do'
+    '(for i in `seq 1 20`; do'
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply tests/test_smoke.py:3034]: Added. Thanks!

File: sky/serve/autoscalers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/serve/autoscalers.py]: just wondering some edge case like the initial delay seconds is about to miss and dependent on network speed 🤔

File: sky/serve/replica_managers.py [hunk 10]
```diff
@@ -235,52 +235,59 @@ class ReplicaStatusProperty:
...
+                # If the service is ever up, we assume there is no bug in the
+                # user code and the scale down is successful, thus enabling the
+                # controller to remove the replica from the replica table and
+                # auto restart the replica.
+                # For replica with a failed sky.launch, it is likely due to some
+                # misconfigured resources, so we don't want to auto restart it.
+                # For replica with a failed sky.down, we cannot restart it since
+                # otherwise we will have a resource leak.
+                return False
+            elif self.first_ready_time < 0:
```
Comment (cblmemo) [inline sky/serve/replica_managers.py:281]:
```suggestion
            else:
```
nit

File: sky/serve/replica_managers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/serve/replica_managers.py:340]: Oh if it is the case, agreed that we keep align w/ spot jobs ;)

**Review by cblmemo — APPROVED** [review]
LGTM! thx :)

File: sky/serve/autoscalers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/serve/autoscalers.py]: Good point! I suppose the latest two loops should guard the concerns : )

File: sky/serve/autoscalers.py [hunk 11]
```diff
@@ -376,12 +392,24 @@ def evaluate_scaling(
...
+                    if info.is_ready:
+                        self.latest_version_ever_ready = self.latest_version
+        for info in latest_replicas:
+            if (info.status_property.unrecoverable_failure() and
+                    self.latest_version_ever_ready < self.latest_version):
+                # Stop scaling if one of replica of the latest version
+                # failed, it is likely that a fatal error happens to the
+                # user application and may lead to a infinte termination
+                # and restart.
+                return []
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/serve/autoscalers.py]: Good catch! Adopted the change!

**Pull request merged**